File: examples/apps/python/msgpacketizer-py/src/msgpacketizerpy/msgpacketizer/decoder.py
import logging

import msgpack

from ...packetizerpy import packetizer
from .common import Message, to_hex


def decode(data: bytes) -> Message | None:
    logging.debug(f"decode input: {to_hex(data)}")
    decoded = packetizer.decode(data)
    if decoded is None:
        return None
    else:
        logging.debug(f"decoded: {to_hex(decoded.data)}")
        msg = msgpack.unpackb(decoded.data, use_list=True, raw=False)  # type: ignore
        logging.debug(f"unpacked: {msg}")
        return Message(decoded.index, msg)

File: examples/apps/go/msgpacketizer-go/pkg/msgpacketizer-go/decoder.go
package msgpacketizer

import (
	// TODO:
	// "github.com/hideakitai/packetizer-go"
	"packetizer-go"

	"github.com/sirupsen/logrus"

	"github.com/vmihailenco/msgpack/v5"
)

func Decode(data []byte) (Message, error) {
	logrus.Debugf("decode input: %s", ToHex(data))
	decoded, err := packetizer.Decode(data)
	if err != nil {
		return Message{}, err
	}
	logrus.Debugf("decoded: %s", ToHex(decoded.Data))
	var msg interface{}
	err = msgpack.Unmarshal(decoded.Data, &msg)
	if err != nil {
		logrus.Warnf("msgpack unmarshal error: %s", err)
		return Message{}, err
	}
	return Message{decoded.Index, msg}, nil
}
